### DISCUSSION TITLE
Optimize CGRAM slots for custom characters

### DISCUSSION THREAD

**Discussion by jlnbotello** [VasilKalchev/LiquidMenu#93] [Ideas]
Hi Vasil,

First of all, thanks for sharing this amazing library. Congrats! 👏 

I've been working with custom characters and I noticed that library use custom characters for left and right focus.
Left and right arrows are already defined on CGROM as 0x7F and 0x7E respectively.
So to release 25% of slots (2/8) of CGRAM, I suggest using CGROM chars.

I hope this could help to improve the library.

Regards,

Julian

Code references:
https://github.com/VaSe7u/LiquidMenu/blob/37dbcb9412725ce78e3b5406cc69e13aaee08779/src/LiquidMenu.cpp#L283
https://github.com/VaSe7u/LiquidMenu/blob/37dbcb9412725ce78e3b5406cc69e13aaee08779/src/LiquidMenu.cpp#L284
https://github.com/VaSe7u/LiquidMenu/blob/1b37ac6b20a84f711d61584ea952bb629a8cf60e/src/LiquidLine.cpp#L113
https://github.com/VaSe7u/LiquidMenu/blob/1b37ac6b20a84f711d61584ea952bb629a8cf60e/src/LiquidLine.cpp#L121

![image](https://user-images.githubusercontent.com/18117253/221438819-e31a9d40-4496-4ca4-bdee-e641fad121f0.png)

Ref: http://www.electronic-engineering.ch/microchip/datasheets/lcd/lcd_data_sheets.html

**Comment by VasilKalchev** [comment]
Hi, @jlnbotello.

Nice catch, I'll give it a test soon. You're welcome to open a pull request if you want.

Thanks for the tip.

**Comment by Jrcarand** [comment]
Hello,
I was going to write exactly the same request. 

I would like to point out though that some "disclaimers" should be added to the library. I found that my LCD is a Hitachi HD44780 and uses ROM A00. People with ROM A02 would have a completely different set (for example I loaded 0b01111110 as right arrow, but in A02 this would be a tilde (~) instead). So, as default A00, but an option to change set should be included.

Here is the reference for the part:
https://www.sparkfun.com/datasheets/LCD/HD44780.pdf

I will try to write a new 'ROMglyphs.h' with names and values and add their behaviour to the LiquidLine class.  Unfortunately I have never used git with other contributors, I can try to make a pull request, this would be my first contribution to a project.

Before I add anything, can I ask if I can create a branch just in case @VasilKalchev ?

**Reply by VasilKalchev** [reply]
Hi @jlnbotello.

Of course, you're welcome to open a pull request. Check out this repo's rough [contributing guide](https://github.com/VasilKalchev/LiquidMenu/blob/master/.github/CONTRIBUTING.md#feature-implementation), you might find [GitHub's guide](https://docs.github.com/en/get-started/exploring-projects-on-github/contributing-to-a-project) more helpful.

I'll be able to test and accept the pull request next week.
